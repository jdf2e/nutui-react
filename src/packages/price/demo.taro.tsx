import React, { useState, useEffect } from 'react'
import Taro from '@tarojs/taro'
import { Cell, Price, CellGroup } from '@/packages/nutui.react.taro'
import { useTranslate } from '@/sites/assets/locale/taro'
import Header from '@/sites/components/header'

interface T {
  title1: string
  title2: string
  title3: string
  title4: string
  title5: string
  title6: string
  title7: string
}

const PriceDemo = () => {
  const [translated] = useTranslate<T>({
    'zh-CN': {
      title1: '支持三种尺寸：small、normal、large',
      title2: '不保留小数',
      title3: '有人民币符号，无千位分隔',
      title4: '有人民币符号，有千位分隔，保留小数点后三位',
      title5: '调整 symbol 符号位置',
      title6: '异步随机变更',
      title7: '不展示 symbol 符号',
    },
    'zh-TW': {
      title1: '支持三種尺寸：small、normal、large',
      title2: '不保留小數',
      title3: '有人民幣符號，無千位分隔',
      title4: '有人民幣符號，有千位分隔，保留小數點後三位',
      title5: '調整 symbol 符號位置',
      title6: '異步隨機變更',
      title7: '不展示 symbol 符號',
    },
    'en-US': {
      title1: 'Support three sizes：small、normal、large',
      title2: 'No decimals',
      title3: 'With RMB symbol, no thousands separator',
      title4:
        'With RMB symbol, separated by thousands, keep three decimal places',
      title5: 'Adjust the symbol position',
      title6: 'Asynchronous random changes',
      title7: 'Do not display symbol',
    },
  })
  const [price, setPrice] = useState(Math.random() * 10000000)

  useEffect(() => {
    const timer = setInterval(() => {
      setPrice(Math.random() * 10000000)
    }, 1000)
    return () => {
      clearInterval(timer)
    }
  }, [])
  return (
    <>
      <Header />
      <div className={`demo ${Taro.getEnv() === 'WEB' ? 'web' : ''}`}>
        <CellGroup title={translated.title1}>
          <Cell>
            <Price price={0} size="small" thousands />
          </Cell>
          <Cell>
            <Price price={0} size="normal" thousands />
          </Cell>
          <Cell>
            <Price price={0} size="large" thousands />
          </Cell>
        </CellGroup>
        <h2>{translated.title2}</h2>
        <Cell>
          <Price price={8888} digits={0} size="normal" thousands />
        </Cell>
        <h2>{translated.title3}</h2>
        <Cell>
          <Price price={10010.01} size="normal" thousands={false} />
        </Cell>
        <h2>{translated.title4}</h2>
        <Cell>
          <Price price={15213.1221} size="normal" digits={3} thousands />
        </Cell>
        <h2>{translated.title5}</h2>
        <Cell>
          <Price
            price={8888.01}
            size="normal"
            position="after"
            symbol="元"
            thousands
          />
        </Cell>
        <h2>{translated.title7}</h2>
        <Cell>
          <Price price={15213.1221} size="normal" symbol="" />
        </Cell>
        <h2>{translated.title6}</h2>
        <Cell>
          <Price price={price} digits={3} size="normal" thousands />
        </Cell>
      </div>
    </>
  )
}

export default PriceDemo
