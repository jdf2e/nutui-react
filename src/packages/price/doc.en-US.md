# Price 

### Intro

It is used to apply different styles to the parts before and after the decimal point of the commodity price value, and also supports functions such as the RMB symbol, thousands separator, and setting the number of decimal points.

### Install

```javascript
import { Price } from '@nutui/nutui-react'
```

## Demo

### Support three sizes：small、normal、large

:::demo

```tsx
import  React from "react"
import { Price, Cell } from '@nutui/nutui-react'

const App = () => {
  return (
<>
        <Cell>
            <Price price={0} size="small" needSymbol thousands />
        </Cell>
        <Cell>
            <Price price={0} size="normal" needSymbol thousands />
        </Cell>
        <Cell>
            <Price price={0} size="large" needSymbol thousands />
        </Cell>
    </>
  );
};
export default App;
```

:::

### No decimals
:::demo

```tsx
import  React from "react"
import { Price, Cell } from '@nutui/nutui-react'

const App = () => {
  return (
    <Cell>
        <Price
          price={8888}
          decimalDigits={0}
          needSymbol
          size="normal"
          thousands
        />
    </Cell>
  );
};
export default App;
```

:::

### With RMB symbol, no thousands separator

:::demo

```tsx
import  React from "react"
import { Price, Cell } from '@nutui/nutui-react'

const App = () => {
  return (
    <Cell>
        <Price price={10010.01} size="normal" needSymbol thousands={false} />
    </Cell>
  );
};
export default App;
```

:::

### With RMB symbol, separated by thousands, keep three decimal places

:::demo

```tsx
import  React from "react"
import { Price, Cell } from '@nutui/nutui-react'

const App = () => {
  return (
    <Cell>
        <Price
          price={15213.1221}
          size="normal"
          decimalDigits={3}
          needSymbol
          thousands
        />
    </Cell>
  );
};
export default App;
```

:::

### Adjust the symbol position

:::demo

```tsx
import  React from "react"
import { Price, Cell } from '@nutui/nutui-react'

const App = () => {
  return (
    <Cell>
        <Price
          price={8888.01}
          size="normal"
          position="after"
          symbol="元"
          needSymbol
          thousands
        />
    </Cell>
  );
};
export default App;
```

:::

### Asynchronous random changes

:::demo

```tsx
import React, { useState, useEffect } from 'react'
import { Price, Cell } from '@nutui/nutui-react'

const App = () => {
  const [price, setPrice] = useState(Math.random() * 10000000)

  useEffect(() => {
    const timer = setInterval(() => {
      setPrice(Math.random() * 10000000)
    }, 1000)
    return () => {
      clearInterval(timer)
    }
  }, [])
  return (
    <Cell>
        <Price
          price={price}
          decimalDigits={3}
          size="normal"
          needSymbol
          thousands
        />
    </Cell>
  );
};
export default App;
```

:::

## API

### Props

| Attribute      | Description                                                | Type            | Default |
|----------------|------------------------------------------------------------|------------------|--------|
| price          | Price                 | Number  | 0      |
| needSymbol    | Add symbol | Boolean | true   |
| symbol         | Symbol type                 | String  | &yen;  |
| decimalDigits | Decimal digits               | Number  | 2      |
| thousands      | Thousands separation   | Boolean | false  |
| position`v1.3.9`   | The symbol appear before or after the price，`before`、`after` | String           | before |
| size`v1.3.9`   | Size，`large`、`normal`、`small`                            | String           | large |
