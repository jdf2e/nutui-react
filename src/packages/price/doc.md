# Price 价格

### 介绍

用来对商品价格数值的小数点前后部分应用不同样式，还支持人民币符号、千位分隔符、设置小数点位数等功能。

### 安装

```javascript
import { Price } from '@nutui/nutui-react'
```

## 代码演示

### 基本用法

:::demo

```tsx
import  React from "react"
import { Price, Cell } from '@nutui/nutui-react'

const App = () => {
  return (
    <Cell>
        <Price price={1010} needSymbol={false} thousands={true} />
    </Cell>
  );
};
export default App;
```

:::

### 有人民币符号，无千位分隔

:::demo

```tsx
import  React from "react"
import { Price, Cell } from '@nutui/nutui-react'

const App = () => {
  return (
    <Cell>
        <Price price={10010.01} needSymbol={true} thousands={false} />
    </Cell>
  );
};
export default App;
```

:::

### 带人民币符号，有千位分隔，保留小数点后三位

:::demo

```tsx
import  React from "react"
import { Price, Cell } from '@nutui/nutui-react'

const App = () => {
  return (
    <Cell>
        <Price price={15213.1221} decimalDigits={3} needSymbol={true} thousands={true} />
    </Cell>
  );
};
export default App;
```

:::

### 异步随机变更

:::demo

```tsx
import React, { useState, useEffect } from 'react'
import { Price, Cell } from '@nutui/nutui-react'

const App = () => {
  const [price, setPrice] = useState(Math.random() * 10000000)

  useEffect(() => {
    const timer = setInterval(() => {
      setPrice(Math.random() * 10000000)
    }, 1000)
    return () => {
      clearInterval(timer)
    }
  }, [])
  return (
    <Cell>
      <Price price={price} decimalDigits={3} needSymbol={true} thousands={true} />
    </Cell>
  );
};
export default App;
```

:::

## API

### Props

| 参数           | 说明                     | 类型    | 默认值 |
| -------------- | ------------------------ | ------- | ------ |
| price          | 价格数量                 | Number  | 0      |
| need-symbol    | 是否需要加上 symbol 符号 | Boolean | true   |
| symbol         | 符号类型                 | String  | &yen;  |
| decimal-digits | 小数位位数               | Number  | 2      |
| thousands      | 是否按照千分号形式显示   | Boolean | false  |
