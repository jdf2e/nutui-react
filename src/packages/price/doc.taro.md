# Price 价格


用来对商品价格数值的小数点前后部分应用不同样式，还支持人民币符号、千位分隔符、设置小数点位数等功能。

## 引入

```tsx
import { Price } from '@nutui/nutui-react-taro'
```

## 示例代码

### 基础用法 small normal large

:::demo

<CodeBlock src='taro/demo1.tsx'></CodeBlock>

:::

### 不保留小数

:::demo

<CodeBlock src='taro/demo2.tsx'></CodeBlock>

:::

### 有人民币符号，无千位分隔

:::demo

<CodeBlock src='taro/demo3.tsx'></CodeBlock>

:::

### 带人民币符号，有千位分隔，保留小数点后三位

:::demo

<CodeBlock src='taro/demo4.tsx'></CodeBlock>

:::

### 调整 symbol 符号位置

:::demo

<CodeBlock src='taro/demo5.tsx'></CodeBlock>

:::

### 不展示 symbol 符号

:::demo

<CodeBlock src='taro/demo6.tsx'></CodeBlock>

:::

### 异步随机变更

:::demo

<CodeBlock src='taro/demo7.tsx'></CodeBlock>

:::

### 划线价

:::demo

<CodeBlock src='taro/demo8.tsx'></CodeBlock>

:::

## Price

### Props

| 属性 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| price | 价格数量 | `number` | `0` |
| symbol | 符号类型 | `string` | `&yen;` |
| digits | 小数位位数 | `number` | `2` |
| thousands | 是否按照千分号形式显示 | `boolean` | `false` |
| position | 符号显示在价格前或者后，`before`、`after` | `string` | `before` |
| size | 价格尺寸，`large`、`normal`、`small` | `string` | `large` |
| line | 是否划线价 | `boolean` | `false` |

## 主题定制

### 样式变量

组件提供了下列 CSS 变量，可用于自定义样式，使用方法请参考 [ConfigProvider 组件](#/zh-CN/component/configprovider)。

| 名称 | 说明 | 默认值 |
| --- | --- | --- |
| \--nutui-price-symbol-big-size | large 尺寸符号字体大小 | `18px` |
| \--nutui-price-integer-big-size | large 尺寸整数部分字体大小 | `24px` |
| \--nutui-price-decimal-big-size | large 尺寸小数部分字体大小 | `18px` |
| \--nutui-price-symbol-medium-size | normal 尺寸符号字体大小 | `14px` |
| \--nutui-price-integer-medium-size | normal 尺寸整数部分字体大小 | `16px` |
| \--nutui-price-decimal-medium-size | normal 尺寸小数部分字体大小 | `14px` |
| \--nutui-price-symbol-small-size | small 尺寸符号字体大小 | `10px` |
| \--nutui-price-integer-small-size | small 尺寸整数部分字体大小 | `12px` |
| \--nutui-price-decimal-small-size | small 尺寸小数部分字体大小 | `10px` |
| \--nutui-price-line-font-size | 划线价字体大小 | `10px` |
| \--nutui-price-line-color | 划线价颜色 |  `#757575` |
| \--nutui-price-symbol-padding-right | 符号的右内边距 |  `1px` |