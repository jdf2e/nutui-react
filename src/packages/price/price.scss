.nut-price {
  font-size: $font-size-3;
  display: inline;
  color: $primary-color;
  &__symbol {
    display: inline-block;
    font-size: $font-size-3;
  }
  &__big {
    display: inline-block;
    font-size: $price-big-size;
  }
  &__point {
    display: inline-block;
    font-size: $price-big-size;
  }
  &__small {
    display: inline-block;
    font-size: $font-size-4;
  }
}
