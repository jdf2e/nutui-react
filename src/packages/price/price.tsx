import React, { FunctionComponent } from 'react'

import bem from '@/utils/bem'

export interface PriceProps {
  price: number | string
  needSymbol: boolean
  symbol: string
  decimalDigits: number
  thousands: boolean
  className: string
  style: React.CSSProperties
}
const defaultProps = {
  price: 0,
  needSymbol: true,
  symbol: '&yen;',
  decimalDigits: 2,
  thousands: false,
  className: '',
} as PriceProps
export const Price: FunctionComponent<Partial<PriceProps>> = (props) => {
  const {
    price,
    needSymbol,
    symbol,
    decimalDigits,
    thousands,
    className,
    ...rest
  } = {
    ...defaultProps,
    ...props,
  }
  const b = bem('price')
  const showSymbol = () => {
    return { __html: (needSymbol ? symbol : '') || '' }
  }
  const checkPoint = (price: string | number) => {
    return String(price).indexOf('.') > 0
  }
  const formatThousands = (num: any) => {
    if (Number(num) == 0) {
      num = 0
    }
    if (checkPoint(num)) {
      num = Number(num).toFixed(decimalDigits)
      num =
        typeof num.split('.') === 'string' ? num.split('.') : num.split('.')[0]
    } else {
      num = num.toString()
    }
    if (thousands) {
      return (num || 0).toString().replace(/(\d)(?=(?:\d{3})+$)/g, '$1,')
    }
    return num
  }
  const formatDecimal = (decimalNum: any) => {
    if (Number(decimalNum) == 0) {
      decimalNum = 0
    }

    if (checkPoint(decimalNum)) {
      decimalNum = Number(decimalNum).toFixed(decimalDigits)
      decimalNum =
        typeof decimalNum.split('.') === 'string' ? 0 : decimalNum.split('.')[1]
    } else {
      decimalNum = 0
    }
    const result = `0.${decimalNum}`
    const resultFixed = Number(result).toFixed(decimalDigits)
    return String(resultFixed).substring(2, resultFixed.length)
  }
  return (
    <div className={`${b()} ${className}`} {...rest}>
      {needSymbol ? (
        <div
          className={`${b('symbol')}`}
          dangerouslySetInnerHTML={showSymbol()}
        />
      ) : null}
      <div className={`${b('big')}`}>{formatThousands(price)}</div>
      <div className={`${b('point')}`}>.</div>
      <div className={`${b('small')}`}>{formatDecimal(price)}</div>
    </div>
  )
}

Price.defaultProps = defaultProps
Price.displayName = 'NutPrice'
