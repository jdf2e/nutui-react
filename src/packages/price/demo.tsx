import React, { useState, useEffect } from 'react'
import { Price } from './price'
import { Cell } from '../cell/cell'

const PriceDemo = () => {
  const [price, setPrice] = useState(Math.random() * 10000000)

  useEffect(() => {
    const timer = setInterval(() => {
      setPrice(Math.random() * 10000000)
    }, 1000)
    return () => {
      clearInterval(timer)
    }
  }, [])
  return (
    <div className="demo">
      <h2>基本用法</h2>
      <Cell>
        <Price price={1010} needSymbol={false} thousands={true} />
      </Cell>
      <h2>有人民币符号，无千位分隔</h2>
      <Cell>
        <Price price={10010.01} needSymbol={true} thousands={false} />
      </Cell>
      <h2>带人民币符号，有千位分隔，保留小数点后三位</h2>
      <Cell>
        <Price price={15213.1221} decimalDigits={3} needSymbol={true} thousands={true} />
      </Cell>
      <h2>异步随机变更</h2>
      <Cell>
        <Price price={price} decimalDigits={3} needSymbol={true} thousands={true} />
      </Cell>
    </div>
  )
}

export default PriceDemo
