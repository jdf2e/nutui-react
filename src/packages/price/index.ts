import { Price } from './price'

export default Price
