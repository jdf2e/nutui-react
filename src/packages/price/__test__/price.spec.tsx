import * as React from 'react'

import { render } from '@testing-library/react'
import '@testing-library/jest-dom'

import { Price } from '../price'

test('props test', () => {
  const { container } = render(
    <Price price={1010} needSymbol={false} thousands />
  )
  expect(container.querySelector('.nut-price__big')?.innerHTML).toBe('1,010')
  expect(container.querySelector('.nut-price__small')?.innerHTML).toBe('00')
  expect(container).toMatchSnapshot()
})

test('props thousands test', () => {
  const { container } = render(
    <Price price={10010.01} needSymbol symbol="$" thousands={false} />
  )
  expect(container.querySelector('.nut-price__big')?.innerHTML).toBe('10010')
  expect(container.querySelector('.nut-price__small')?.innerHTML).toBe('01')
  expect(container).toMatchSnapshot()
})

test('props needSymbol and symbol test', () => {
  const { container } = render(<Price price={10010} needSymbol symbol="$" />)
  expect(container.querySelector('.nut-price__symbol')).toBeInTheDocument()
  expect(container.querySelector('.nut-price__symbol')?.innerHTML).toBe('$')
  expect(container).toMatchSnapshot()
})

test('props decimalDigits test', () => {
  const { container } = render(<Price price={15213.1221} decimalDigits={3} />)
  expect(container.querySelector('.nut-price__small')?.innerHTML).toBe('122')
  expect(container).toMatchSnapshot()
})
