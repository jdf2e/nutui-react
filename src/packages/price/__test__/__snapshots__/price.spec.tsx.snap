// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`props digits test 1`] = `
<div>
  <div
    class="nut-price  "
  >
    <div
      class="nut-price-symbol nut-price-symbol-large"
    >
      Â¥
    </div>
    <div
      class="nut-price-integer nut-price-integer-large "
    >
      15213
    </div>
    <div
      class="nut-price-decimal nut-price-decimal-large "
    >
      .
    </div>
    <div
      class="nut-price-decimal nut-price-decimal-large "
    >
      122
    </div>
  </div>
</div>
`;

exports[`props symbol test 1`] = `
<div>
  <div
    class="nut-price  "
  >
    <div
      class="nut-price-symbol nut-price-symbol-large"
    >
      $
    </div>
    <div
      class="nut-price-integer nut-price-integer-large "
    >
      10010
    </div>
    <div
      class="nut-price-decimal nut-price-decimal-large "
    >
      .
    </div>
    <div
      class="nut-price-decimal nut-price-decimal-large "
    >
      00
    </div>
  </div>
</div>
`;

exports[`props test 1`] = `
<div>
  <div
    class="nut-price  "
  >
    <div
      class="nut-price-integer nut-price-integer-large "
    >
      1,010
    </div>
    <div
      class="nut-price-decimal nut-price-decimal-large "
    >
      .
    </div>
    <div
      class="nut-price-decimal nut-price-decimal-large "
    >
      00
    </div>
  </div>
</div>
`;

exports[`props thousands test 1`] = `
<div>
  <div
    class="nut-price  "
  >
    <div
      class="nut-price-symbol nut-price-symbol-large"
    >
      $
    </div>
    <div
      class="nut-price-integer nut-price-integer-large "
    >
      10010
    </div>
    <div
      class="nut-price-decimal nut-price-decimal-large "
    >
      .
    </div>
    <div
      class="nut-price-decimal nut-price-decimal-large "
    >
      01
    </div>
  </div>
</div>
`;
